File: tui/src/session-state.ts
import type {ProtocolEvent} from './protocol.js';

/** Mocked-session event accepted by the local conversation projection. */
export type SessionEvent = Extract<
  ProtocolEvent,
  {
    readonly type:
      | 'session.started'
      | 'assistant.delta'
      | 'assistant.completed'
      | 'session.completed'
      | 'session.cancelled';
  }
>;

/** One local command or validated event in the ordered session projection stream. */
export type SessionUpdate =
  | {
      readonly type: 'task.submitted';
      readonly commandId: string;
      readonly task: string;
    }
  | {
      readonly type: 'cancel.requested';
      readonly commandId: string;
      readonly sessionId: string;
    }
  | {readonly type: 'event.received'; readonly event: SessionEvent};

/** Visible state for one submitted task and its authoritative Python session. */
export interface ConversationTurn {
  /** Command that caused this turn. */
  readonly commandId: string;
  /** Exact user text sent in the `session.start` payload. */
  readonly task: string;
  /** Python-owned session identity, available after `session.started`. */
  readonly sessionId?: string;
  /** Session-local lifecycle projected from validated events. */
  readonly status: 'starting' | 'running' | 'cancelling' | 'completed' | 'cancelled';
  /** Accepted assistant deltas in sequence order. */
  readonly assistantText: string;
  /** Last accepted Python sequence, or zero before the session starts. */
  readonly lastSequence: number;
  /** Whether Python has confirmed the accumulated assistant text. */
  readonly assistantCompleted: boolean;
  /** Command that requested cancellation, present only after a validated local request. */
  readonly cancelCommandId?: string;
}

/**
 * Local, immutable conversation projection rendered by Ink.
 *
 * `idle` has no turns. A local submission enters `starting`; only `session.started` enters
 * `running`, and only a terminal Python event enters a terminal state. `protocol-failed` is a local
 * fail-closed state for an event that violates correlation, sequence, identity, or completion
 * invariants. Turns remain visible after runtime exit because runtime lifecycle changes are reduced
 * separately.
 */
export interface SessionState {
  /** Status of the newest turn, or `idle` before the first submission. */
  readonly status:
    | 'idle'
    | 'starting'
    | 'running'
    | 'cancelling'
    | 'completed'
    | 'cancelled'
    | 'protocol-failed';
  /** Completed turns plus the optional active turn in submission order. */
  readonly turns: readonly ConversationTurn[];
  /** Safe invariant failure displayed when an event cannot enter trusted UI state. */
  readonly protocolFailure?: string;
}

/** Empty projection used when the TUI first mounts. */
export const INITIAL_SESSION_STATE: SessionState = {status: 'idle', turns: []};

/**
 * Reduce one ordered local/session update into visible conversation state.
 *
 * The reducer is pure. It never guesses around duplicate, out-of-order, mismatched, or unknown
 * transitions: the first violation produces `protocol-failed`, and later updates are ignored.
 * A local cancellation request enters `cancelling`, but only Python's `session.cancelled` event
 * enters the terminal `cancelled` state. Start-correlated stream and completion events remain legal
 * while cancellation is pending, so whichever valid terminal event arrives first wins.
 * `assistant.completed` must exactly equal accepted deltas, and `session.completed` is legal only
 * after that confirmation. Runtime lifecycle changes are intentionally outside this reducer.
 *
 * @param state - Current immutable conversation projection.
 * @param update - Local submission or validated session event in arrival order.
 * @returns A new projection, or the existing terminal protocol-failure projection.
 */
export function reduceSessionState(state: SessionState, update: SessionUpdate): SessionState {
  if (state.status === 'protocol-failed') {
    return state;
  }
  if (update.type === 'task.submitted') {
    return reduceSubmission(state, update);
  }
  if (update.type === 'cancel.requested') {
    return reduceCancellationRequest(state, update);
  }
  return reduceEvent(state, update.event);
}

function reduceSubmission(
  state: SessionState,
  update: Extract<SessionUpdate, {readonly type: 'task.submitted'}>,
): SessionState {
  if (
    state.status === 'starting' ||
    state.status === 'running' ||
    state.status === 'cancelling'
  ) {
    return failProjection(state, 'A second task was submitted while a session was active.');
  }
  const turn: ConversationTurn = {
    commandId: update.commandId,
    task: update.task,
    status: 'starting',
    assistantText: '',
    lastSequence: 0,
    assistantCompleted: false,
  };
  return {status: 'starting', turns: [...state.turns, turn]};
}

function reduceCancellationRequest(
  state: SessionState,
  update: Extract<SessionUpdate, {readonly type: 'cancel.requested'}>,
): SessionState {
  const turn = state.turns.at(-1);
  if (
    state.status !== 'running' ||
    turn?.status !== 'running' ||
    turn.sessionId === undefined
  ) {
    return failProjection(state, 'Cancellation was requested without an addressable active session.');
  }
  if (update.sessionId !== turn.sessionId) {
    return failProjection(state, 'Cancellation did not target the active session.');
  }
  return replaceNewestTurn(
    state,
    {...turn, status: 'cancelling', cancelCommandId: update.commandId},
    'cancelling',
  );
}

function reduceEvent(state: SessionState, event: SessionEvent): SessionState {
  const turn = state.turns.at(-1);
  if (
    turn === undefined ||
    (turn.status !== 'starting' && turn.status !== 'running' && turn.status !== 'cancelling')
  ) {
    return failProjection(state, `Received ${event.type} without an active submitted task.`);
  }
  const expectedCorrelationId =
    event.type === 'session.cancelled' ? turn.cancelCommandId : turn.commandId;
  if (event.correlation_id !== expectedCorrelationId) {
    return failProjection(state, `${event.type} did not correlate to the active task.`);
  }

  if (event.type === 'session.started') {
    if (turn.status !== 'starting' || event.sequence !== 1) {
      return failProjection(state, 'session.started must be the first event for a submitted task.');
    }
    return replaceNewestTurn(state, {
      ...turn,
      sessionId: event.session_id,
      status: 'running',
      lastSequence: event.sequence,
    });
  }

  const mismatch = sessionEventMismatch(turn, event);
  if (mismatch !== undefined) {
    return failProjection(state, mismatch);
  }

  switch (event.type) {
    case 'assistant.delta':
      if (turn.assistantCompleted) {
        return failProjection(state, 'assistant.delta arrived after assistant completion.');
      }
      return replaceNewestTurn(state, {
        ...turn,
        assistantText: turn.assistantText + event.payload.text,
        lastSequence: event.sequence,
      });
    case 'assistant.completed':
      if (turn.assistantCompleted || event.payload.text !== turn.assistantText) {
        return failProjection(
          state,
          'assistant.completed did not exactly confirm the accumulated deltas.',
        );
      }
      return replaceNewestTurn(state, {
        ...turn,
        assistantCompleted: true,
        lastSequence: event.sequence,
      });
    case 'session.completed':
      if (!turn.assistantCompleted) {
        return failProjection(state, 'session.completed arrived before assistant completion.');
      }
      return replaceNewestTurn(
        state,
        {...turn, status: 'completed', lastSequence: event.sequence},
        'completed',
      );
    case 'session.cancelled':
      if (turn.status !== 'cancelling' || turn.cancelCommandId === undefined) {
        return failProjection(state, 'session.cancelled arrived without a cancellation request.');
      }
      return replaceNewestTurn(
        state,
        {...turn, status: 'cancelled', lastSequence: event.sequence},
        'cancelled',
      );
  }
}

function sessionEventMismatch(
  turn: ConversationTurn,
  event: Exclude<SessionEvent, {readonly type: 'session.started'}>,
): string | undefined {
  if (
    (turn.status !== 'running' && turn.status !== 'cancelling') ||
    turn.sessionId === undefined
  ) {
    return `${event.type} arrived before session.started.`;
  }
  if (event.session_id !== turn.sessionId) {
    return `${event.type} did not belong to the active session.`;
  }
  if (event.sequence !== turn.lastSequence + 1) {
    return `${event.type} did not carry the next session sequence.`;
  }
  return undefined;
}

function replaceNewestTurn(
  state: SessionState,
  turn: ConversationTurn,
  status: SessionState['status'] = turn.status === 'cancelling' ? 'cancelling' : 'running',
): SessionState {
  return {status, turns: [...state.turns.slice(0, -1), turn]};
}

function failProjection(state: SessionState, message: string): SessionState {
  return {...state, status: 'protocol-failed', protocolFailure: message};
}

File: src/code_assist_harness/session_state.py
"""Pure, replayable lifecycle state for one harness session."""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass, replace
from typing import Literal

from .protocol import (
    AssistantCompletedEvent,
    AssistantDeltaEvent,
    SessionCancelledEvent,
    SessionCompletedEvent,
    SessionEvent,
    SessionFailedEvent,
    SessionStartedEvent,
)

type SessionStatus = Literal[
    "idle",
    "starting",
    "running",
    "awaiting_approval",
    "cancelling",
    "completed",
    "cancelled",
    "failed",
]
"""Lifecycle states owned by the harness for one session."""

type InvariantFailureCode = Literal[
    "illegal_transition",
    "terminal_state_absorbing",
    "correlation_mismatch",
    "session_mismatch",
    "sequence_gap",
    "sequence_regression",
    "assistant_after_completion",
    "assistant_already_completed",
    "assistant_completion_mismatch",
    "session_completion_before_assistant",
]
"""Payload-free reasons that an update could not enter trusted state."""

TERMINAL_SESSION_STATUSES = frozenset[SessionStatus]({"completed", "cancelled", "failed"})
"""States that cannot accept another update for the same session."""


@dataclass(frozen=True, slots=True)
class SessionFailure:
    """Safe terminal failure copied from a validated ``session.failed`` event."""

    code: str
    message: str


@dataclass(frozen=True, slots=True)
class SessionState:
    """Immutable authoritative projection for exactly one session.

    The reducer creates every non-idle value. ``last_sequence`` advances only after an accepted
    validated wire event; command-originated domain facts do not consume protocol sequence values.
    Terminal states are absorbing, so a conversation owner must create a fresh state for a later
    task rather than reusing this object.
    """

    status: SessionStatus = "idle"
    start_command_id: str | None = None
    task: str | None = None
    session_id: str | None = None
    cancel_command_id: str | None = None
    last_sequence: int = 0
    assistant_text: str = ""
    assistant_completed: bool = False
    session_failure: SessionFailure | None = None


INITIAL_SESSION_STATE = SessionState()
"""Canonical empty state used before a task is submitted."""


@dataclass(frozen=True, slots=True)
class TaskSubmitted:
    """Trusted domain fact recording a local task submission."""

    command_id: str
    task: str


@dataclass(frozen=True, slots=True)
class CancelRequested:
    """Trusted domain fact recording a cancellation sent for the active session."""

    command_id: str
    session_id: str


@dataclass(frozen=True, slots=True)
class ApprovalRequested:
    """Trusted domain fact recording that the active session is waiting for approval.

    This is a harness-owned lifecycle fact, not a protocol-v1 wire message. A later unit can attach
    approval request details without expanding the reducer's lifecycle vocabulary.
    """

    session_id: str


@dataclass(frozen=True, slots=True)
class ApprovalResolved:
    """Trusted domain fact recording that an approval wait has ended."""

    session_id: str


type SessionUpdate = (
    TaskSubmitted | CancelRequested | ApprovalRequested | ApprovalResolved | SessionEvent
)
"""One validated fact accepted by the session reducer."""


@dataclass(frozen=True, slots=True)
class SessionInvariantFailure:
    """Safe explanation for a rejected update.

    The diagnostic deliberately excludes command IDs, session IDs, task text, assistant text, and
    event payloads. That makes it suitable for a transcript or user-visible status without turning
    an invariant failure into a secret-exfiltration path.
    """

    code: InvariantFailureCode
    prior_status: SessionStatus
    event_type: str


@dataclass(frozen=True, slots=True)
class SessionReduction:
    """Result of reducing one update or replaying an ordered update stream.

    ``failure`` is absent for an accepted reduction. When it is present, ``state`` is the exact
    prior object supplied to the failing reduction.
    """

    state: SessionState
    failure: SessionInvariantFailure | None = None

    @property
    def ok(self) -> bool:
        """Return whether every attempted update was accepted."""
        return self.failure is None


def reduce_session_state(state: SessionState, update: SessionUpdate) -> SessionReduction:
    """Reduce one trusted domain fact or validated session event without side effects.

    Correlation, session identity, and contiguous sequence checks run before a legal wire event can
    alter the projection. The reducer performs no I/O, clock reads, randomness, mutation, provider
    work, or protocol parsing. Callers must validate raw wire values before constructing this API's
    ``SessionEvent`` input.

    Args:
        state: Current immutable state for one session.
        update: One trusted local fact or validated protocol-v1 session event.

    Returns:
        The accepted next state, or the exact prior state plus a payload-free invariant failure.
    """
    event_type = _event_type(update)
    if state.status in TERMINAL_SESSION_STATUSES:
        return _reject(state, "terminal_state_absorbing", event_type)

    if isinstance(update, TaskSubmitted):
        if state.status != "idle":
            return _reject(state, "illegal_transition", event_type)
        return _accept(
            SessionState(
                status="starting",
                start_command_id=update.command_id,
                task=update.task,
            )
        )

    if isinstance(update, CancelRequested):
        if state.status not in {"running", "awaiting_approval"}:
            return _reject(state, "illegal_transition", event_type)
        if update.session_id != state.session_id:
            return _reject(state, "session_mismatch", event_type)
        return _accept(replace(state, status="cancelling", cancel_command_id=update.command_id))

    if isinstance(update, ApprovalRequested):
        if state.status != "running":
            return _reject(state, "illegal_transition", event_type)
        if update.session_id != state.session_id:
            return _reject(state, "session_mismatch", event_type)
        return _accept(replace(state, status="awaiting_approval"))

    if isinstance(update, ApprovalResolved):
        if state.status != "awaiting_approval":
            return _reject(state, "illegal_transition", event_type)
        if update.session_id != state.session_id:
            return _reject(state, "session_mismatch", event_type)
        return _accept(replace(state, status="running"))

    return _reduce_session_event(state, update)


def replay_session_updates(
    updates: Iterable[SessionUpdate],
    initial_state: SessionState = INITIAL_SESSION_STATE,
) -> SessionReduction:
    """Replay an ordered update stream into the same result on every invocation.

    Replay stops at the first invariant failure because that update and every later update are
    outside the trusted projection. It never resumes work or re-executes a side effect.

    Args:
        updates: Ordered trusted facts and validated session events.
        initial_state: Immutable state from which replay begins.

    Returns:
        The final accepted state, or the state immediately before the first rejected update and its
        safe failure.
    """
    state = initial_state
    for update in updates:
        result = reduce_session_state(state, update)
        if not result.ok:
            return result
        state = result.state
    return _accept(state)


def _reduce_session_event(state: SessionState, event: SessionEvent) -> SessionReduction:
    event_type = event.type
    legal_statuses: dict[str, frozenset[SessionStatus]] = {
        "session.started": frozenset({"starting"}),
        "assistant.delta": frozenset({"running", "cancelling"}),
        "assistant.completed": frozenset({"running", "cancelling"}),
        "session.completed": frozenset({"running", "cancelling"}),
        "session.cancelled": frozenset({"cancelling"}),
        "session.failed": frozenset({"running", "awaiting_approval", "cancelling"}),
    }
    if state.status not in legal_statuses[event_type]:
        return _reject(state, "illegal_transition", event_type)

    expected_correlation = (
        state.cancel_command_id
        if isinstance(event, SessionCancelledEvent)
        else state.start_command_id
    )
    if event.correlation_id != expected_correlation:
        return _reject(state, "correlation_mismatch", event_type)

    if isinstance(event, SessionStartedEvent):
        sequence_failure = _sequence_failure(state, event.sequence, event_type)
        if sequence_failure is not None:
            return sequence_failure
        return _accept(
            replace(
                state,
                status="running",
                session_id=event.session_id,
                last_sequence=event.sequence,
            )
        )

    if event.session_id != state.session_id:
        return _reject(state, "session_mismatch", event_type)

    sequence_failure = _sequence_failure(state, event.sequence, event_type)
    if sequence_failure is not None:
        return sequence_failure

    if isinstance(event, AssistantDeltaEvent):
        if state.assistant_completed:
            return _reject(state, "assistant_after_completion", event_type)
        return _accept(
            replace(
                state,
                last_sequence=event.sequence,
                assistant_text=state.assistant_text + event.payload.text,
            )
        )

    if isinstance(event, AssistantCompletedEvent):
        if state.assistant_completed:
            return _reject(state, "assistant_already_completed", event_type)
        if event.payload.text != state.assistant_text:
            return _reject(state, "assistant_completion_mismatch", event_type)
        return _accept(replace(state, last_sequence=event.sequence, assistant_completed=True))

    if isinstance(event, SessionCompletedEvent):
        if not state.assistant_completed:
            return _reject(state, "session_completion_before_assistant", event_type)
        return _accept(replace(state, status="completed", last_sequence=event.sequence))

    if isinstance(event, SessionCancelledEvent):
        return _accept(replace(state, status="cancelled", last_sequence=event.sequence))

    if isinstance(event, SessionFailedEvent):
        return _accept(
            replace(
                state,
                status="failed",
                last_sequence=event.sequence,
                session_failure=SessionFailure(
                    code=event.payload.code,
                    message=event.payload.message,
                ),
            )
        )

    raise AssertionError(f"unhandled validated session event: {event_type}")


def _sequence_failure(
    state: SessionState, sequence: int, event_type: str
) -> SessionReduction | None:
    expected_sequence = state.last_sequence + 1
    if sequence < expected_sequence:
        return _reject(state, "sequence_regression", event_type)
    if sequence > expected_sequence:
        return _reject(state, "sequence_gap", event_type)
    return None


def _accept(state: SessionState) -> SessionReduction:
    return SessionReduction(state=state)


def _reject(state: SessionState, code: InvariantFailureCode, event_type: str) -> SessionReduction:
    return SessionReduction(
        state=state,
        failure=SessionInvariantFailure(
            code=code,
            prior_status=state.status,
            event_type=event_type,
        ),
    )


def _event_type(update: SessionUpdate) -> str:
    if isinstance(update, TaskSubmitted):
        return "task.submitted"
    if isinstance(update, CancelRequested):
        return "cancel.requested"
    if isinstance(update, ApprovalRequested):
        return "approval.requested"
    if isinstance(update, ApprovalResolved):
        return "approval.resolved"
    return update.type


__all__ = [
    "INITIAL_SESSION_STATE",
    "TERMINAL_SESSION_STATUSES",
    "ApprovalRequested",
    "ApprovalResolved",
    "CancelRequested",
    "InvariantFailureCode",
    "SessionFailure",
    "SessionInvariantFailure",
    "SessionReduction",
    "SessionState",
    "SessionStatus",
    "SessionUpdate",
    "TaskSubmitted",
    "reduce_session_state",
    "replay_session_updates",
]
